rmw_cyclonedds_cpp/src/fallthrough_macro.hpp

```diff
@@ -15,15 +15,18 @@
 #ifndef FALLTHROUGH_MACRO_HPP_
 #define FALLTHROUGH_MACRO_HPP_
 
-#if __has_cpp_attribute(fallthrough) || (__cplusplus >= 201603L)
+#if __has_cpp_attribute(fallthrough) && (__cplusplus >= 201703L)
 // C++17
 #define FALLTHROUGH [[fallthrough]]
 #elif __has_cpp_attribute(clang::fallthrough)
 // Clang
 #define FALLTHROUGH [[clang::fallthrough]]
+#elif __has_cpp_attribute(gnu::fallthrough)
+// gcc with gnu extension
+#define FALLTHROUGH [[gnu::fallthrough]]
 #else
 // gcc
-#define FALLTHROUGH /* fallthrough */
+#define FALLTHROUGH __attribute__ ((fallthrough));
```

Comment:
The `/* fallthrough */` comment actually *does* suppress the warning on GCC, so the `[[gnu::fallthrough]]` case is unneeded and redundant. If you prefer `[[gnu::fallthrough]]` for GCC we can definitely keep it, as it does lend credibility to the code. The default case should not be `__attribute__((fallthrough))`, as you've discovered.

Reply:
Actually, yeah! I like the `[[gnu::fallthrough]]` since it works with `-Wimplicit-fallthrough=5` https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc.pdf